usr.sbin/bhyve/block_if.h

```diff
@@ -57,12 +80,52 @@ struct blockif_req {
 	void		*br_param;
 	struct iovec	br_iov[BLOCKIF_IOV_MAX];
 };
+typedef struct blockif_req blockif_req_t;
+
+struct blockif_elem {
+        TAILQ_ENTRY(blockif_elem) be_link;
+        struct blockif_req  *be_req;
+        enum blockop         be_op;
+        enum blockstat       be_status;
+        pthread_t            be_tid;
+        off_t                be_block;
+};
+typedef struct blockif_elem blockif_elem_t;
 
-struct blockif_ctxt;
-struct blockif_ctxt *blockif_open(const char *optstr, const char *ident);
+typedef struct block_backend block_backend_t;
+struct blockif_ctxt {
+        int                     bc_magic;
+        union {
+                int             bc_fd;
+        	/* For data specific for this instance of the backend*/
+	        void           *bc_desc;
+	} desc;
```

Comment:
Every slot in in the PCI-table needs a to be connected to a driver....
So IMHO the a slot is connected to backend instance that goes with the data of that PCI slot.
For example 
 - every file: slot needs its own backend struc to register the filedescriptor.
 - every rbd: slot needs its onw backend to point to an rbd descriptor.
Other things that will differ per PCI-slot is the size.

Then that backend also points to the backend driver that it is using, like file: or rbd:

I've made it a pointer, since most of the backend_descriptors will be a allocated structure that is usually more that just an integer.

Reply:
Ok, I forgot what `bb_opaque` was for. Anyway, I don't think we need a union, just a `void *` pointer, so every driver will cast it to its own struct or whatever it wants.